libbeat/statestore/backend/memlog/diskstore.go

```diff
@@ -165,5 +264,410 @@ func (s *diskstore) LogOperation(op op) error {
+	tempfile := baseName + ".new"
+	f, err := os.OpenFile(tempfile, os.O_RDWR|os.O_CREATE|os.O_TRUNC|os.O_SYNC, s.fileMode)
+	if err != nil {
+		return "", err
+	}
+
+	ok := false
+	defer cleanup.IfNot(&ok, func() {
+		f.Close()
+	})
+
+	writer := bufio.NewWriterSize(&ensureWriter{f}, s.bufferSize)
+	enc := newJSONEncoder(writer)
+	if _, err = writer.Write([]byte{'['}); err != nil {
+		return "", err
+	}
+
+	first := true
+	for key, entry := range states {
+		prefix := []byte(",\n")
+		if first {
+			prefix = prefix[1:]
+			first = false
+		}
+		if _, err = writer.Write(prefix); err != nil {
+			return "", err
+		}
+
+		err = enc.Encode(storeEntry{
+			Key:    key,
+			Fields: entry.value,
+		})
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if _, err = writer.Write([]byte("\n]")); err != nil {
+		return "", err
+	}
+
+	if err = writer.Flush(); err != nil {
+		return "", err
+	}
+
+	if err = syncFile(f); err != nil {
+		return "", err
+	}
+
+	ok = true
+	if err = f.Close(); err != nil {
+		return "", err
+	}
+
+	return tempfile, nil
+}
+
+func (s *diskstore) checkpointClearLog() {
+	if s.logFile == nil {
+		s.logNeedsTruncate = true
+		return
+	}
+
+	err := s.logFile.Truncate(0)
+	if err == nil {
+		_, err = s.logFile.Seek(0, os.SEEK_SET)
+	}
+
+	if err != nil {
+		s.logFile.Close()
+		s.logFile = nil
+		s.logBuf = nil
+		s.logNeedsTruncate = true
+		s.logInvalid = true
+	}
+
+	s.logEntries = 0
+	s.logFileSize = 0
+}
+
+func updateActiveMarker(log *logp.Logger, home, active string) error {
+	activeLink := filepath.Join(home, "active.dat")
+	tmpLink := filepath.Join(home, "active.dat")
+	log = log.With("temporary", tmpLink, "data_file", active, "link_file", activeLink)
+
+	if active == "" {
+		if err := os.Remove(activeLink); err != nil { // try, remove active symlink if present.
+			log.Errorf("Failed to remove old pointer file: %v", err)
+		}
+		return nil
+	}
+
+	// Atomically try to update the pointer file to the most recent data file.
+	// We 'simulate' the atomic update by create the temporary active.json.tmp symlink file,
+	// which we rename to active.json. If active.json.tmp exists we remove it.
+	if err := os.Remove(tmpLink); err != nil && !os.IsNotExist(err) {
+		log.Errorf("Failed to remove old temporary symlink file: %v", err)
+		return err
+	}
+	if err := ioutil.WriteFile(tmpLink, []byte(active), 0600); err != nil {
+		log.Errorf("Failed to write temporary pointer file: %v", err)
+		return err
+	}
+	if err := os.Rename(tmpLink, activeLink); err != nil {
+		log.Errorf("Failed to replace link file: %v", err)
+		return err
+	}
+
+	trySyncPath(home)
+	return nil
+}
+
+// removeOldDataFiles sorts the data files by their update sequence number and
+// finally deletes all but the newest file from the storage directory.
+func (s *diskstore) removeOldDataFiles() {
+	L := len(s.dataFiles)
+	if L <= 1 {
+		return
+	}
+
+	removable, keep := s.dataFiles[:L-1], s.dataFiles[L-1:]
```

Comment:
It looks like `keep` will always be length 1 in this function, can we just make it a `dataFileInfo` and assign it from `s.dataFiles[L-1]` instead?

Reply:
It is correct that the length is 1. We advance through the slice by finally assigning (assuming we haven't had an error) `s.dataFiles = keep`. If `len(dataFiles) == 1`, then this operation becomes a noop.
if `len(dataFiles) > 1`, then the active data file is the last one, while the other ones are all considered old. I think I will rather split the slice into two variables: `oldDataFiles []dataFileInfo` and `activeDataFile dataFileInfo`. Then it is more clear which is which and we can always attempt to remove all entries in `oldDataFiles`.